pkg/schema/type_check_test.go

```diff
@@ -0,0 +1,194 @@
+
+	"github.com/authzed/spicedb/pkg/schemadsl/compiler"
+)
+
+func TestTypechecking(t *testing.T) {
+	t.Parallel()
+	type testcase struct {
+		name       string
+		schemaText string
+		expected   map[string][]string
+	}
+	tcs := []testcase{
+		{
+			name: "basic arrow",
+			schemaText: `
+			definition user {}
+
+			definition organization {
+				relation member: user
+			}
+
+			definition resource {
+				relation org: organization
+				relation viewer: user
+				permission view = org->member + viewer
+			}
+		`,
+			expected: map[string][]string{
+				"organization#member": {"user"},
+				"resource#viewer":     {"user"},
+				"resource#org":        {"organization"},
+				"resource#view":       {"user"},
+			},
+		},
+		{
+			name: "multi-type arrow",
+			schemaText: `
+			definition user {}
+
+			definition organization {
+				relation member: user
+			}
+
+			definition resource {
+				relation org: organization
+				relation viewer: user
+				permission view = org + viewer
+			}
+		`,
+			expected: map[string][]string{
+				"organization#member": {"user"},
+				"resource#viewer":     {"user"},
+				"resource#org":        {"organization"},
+				"resource#view":       {"organization", "user"},
+			},
+		},
+		{
+			name: "functional",
+			schemaText: `
+			definition user {}
+
+			definition organization {
+				relation member: user
+			}
+
+			definition resource {
+				relation org: organization
+				permission view = org.all(member)
+			}
+		`,
+			expected: map[string][]string{
+				"organization#member": {"user"},
+				"resource#viewer":     {"user"},
+				"resource#org":        {"organization"},
+				"resource#view":       {"user"},
+			},
+		},
+		{
+			name: "multi-type rel",
+			schemaText: `
+			definition user {}
+
+			definition organization {
+				relation member: user
+			}
+
+			definition resource {
+				relation viewer: user | organization
+			}
+		`,
+			expected: map[string][]string{
+				"organization#member": {"user"},
+				"resource#viewer":     {"user", "organization"},
+			},
+		},
+		{
+			name: "subrel",
+			schemaText: `
+			definition user {}
+
+			definition organization {
+				relation member: user
+			}
+
+			definition resource {
+				relation viewer: organization#member
+			}
+		`,
+			expected: map[string][]string{
+				"organization#member": {"user"},
+				"resource#viewer":     {"user"},
+			},
+		},
+		{
+			name: "wildcard",
+			schemaText: `
+			definition user {}
+
+			definition organization {
+				relation member: user:*
+			}
+
+			definition resource {
+				relation viewer: organization#member
+			}
+		`,
+			expected: map[string][]string{
+				"organization#member": {"user"},
+				"resource#viewer":     {"user"},
+			},
+		},
+		{
+			name: "banned",
+			schemaText: `
+			definition user {}
+
+			definition organization {
+				relation member: user
+			}
+
+			definition resource {
+				relation viewer: organization#member
+				relation banned: user
+				permission view = viewer - banned
```

Comment:
So this raises an interesting question: the target subject types of `view` here are *technically* `user` *and* `organization#member`, since you can issue a check for organization#member here. Question is: do we include that in this set?

Reply:
I think we do include it and make it required to be specified

Reply:
`organization#member` is also a `user` though (or, really, the set of things that are on the other end of that relation -- and here we have computed subtypes again). I'm okay with it being specified, though

Reply:
Yeah, but its valid, so I think we should add it